internal/scheduler/legacyscheduler.go

```diff
@@ -95,595 +97,579 @@ func SchedulingConstraintsFromSchedulingConfig(
+		return nil, err
+	}
+	for _, job := range rescheduledJobs {
+		if _, ok := preemptedJobsById[job.GetId()]; ok {
+			delete(preemptedJobsById, job.GetId())
+		} else {
+			scheduledJobsById[job.GetId()] = job
 		}
-		gangId, gangCardinality, isGangJob, err := GangIdAndCardinalityFromAnnotations(
-			job.GetAnnotations(),
-			it.gangIdAnnotation,
-			it.gangCardinalityAnnotation,
+	}
+
+	// Evict jobs on oversubscribed nodes.
+	evictorResult, inMemoryJobRepo, err = sch.evict(
+		ctxlogrus.ToContext(
+			ctx,
+			log.WithField("stage", "evict oversubscribed"),
+		),
+		NewOversubscribedEvictor(
+			sch.jobRepo,
+			sch.config.Preemption.PriorityClasses,
+			sch.nodeOversubscribedEvictionProbability,
+		),
+	)
+	if err != nil {
+		return nil, err
+	}
+	maps.Copy(preemptedJobsById, evictorResult.EvictedJobsById)
+
+	// Re-schedule evicted jobs/schedule new jobs.
+	// Only necessary if a non-zero number of jobs were evicted.
+	if len(evictorResult.EvictedJobsById) > 0 {
+		rescheduledJobs, err = sch.schedule(
+			ctxlogrus.ToContext(
+				ctx,
+				log.WithField("stage", "reschedule after oversubscribed eviction"),
+			),
+			inMemoryJobRepo,
+			// Only evicted jobs should be scheduled in this round,
+			// so we set an empty repo for queued jobs.
+			NewInMemoryJobRepository(sch.nodeDb.priorityClasses),
 		)
 		if err != nil {
-			log := ctxlogrus.Extract(it.ctx)
-			logging.WithStacktrace(log, err).Errorf("failed to get gang cardinality for job %s", job.GetId())
-			gangCardinality = 1 // Schedule jobs with invalid gang cardinality one by one.
+			return nil, err
 		}
-		if isGangJob {
-			it.jobsByGangId[gangId] = append(it.jobsByGangId[gangId], job)
-			gang := it.jobsByGangId[gangId]
-			if len(gang) == gangCardinality {
-				delete(it.jobsByGangId, gangId)
-				it.next = gang
-				return it.next, nil
+		for _, job := range rescheduledJobs {
+			if _, ok := preemptedJobsById[job.GetId()]; ok {
+				delete(preemptedJobsById, job.GetId())
+			} else {
+				scheduledJobsById[job.GetId()] = job
 			}
-		} else {
-			it.next = []LegacySchedulerJob{job}
-			return it.next, nil
 		}
 	}
-}
 
-func (it *QueuedGangIterator) hitLookbackLimit() bool {
-	if it.maxLookback == 0 {
-		return false
+	nodeByJobId, err := sch.validateSchedulingConsistency(
+		ctxlogrus.ToContext(
+			ctx,
+			log.WithField("stage", "validate consistency"),
+		),
+		snapshot,
+		preemptedJobsById,
+		scheduledJobsById,
+	)
+	if err != nil {
+		return nil, err
 	}
-	return it.jobsSeen > it.maxLookback
-}
 
-// QueueCandidateGangIterator is an iterator over gangs in a queue that could be scheduled
-// without exceeding per-queue limits.
-type QueueCandidateGangIterator struct {
-	ctx context.Context
-	SchedulingConstraints
-	QueueSchedulingRoundReport *QueueSchedulingRoundReport
-	queuedGangIterator         *QueuedGangIterator
+	preemptedJobs := maps.Values(preemptedJobsById)
+	scheduledJobs := maps.Values(scheduledJobsById)
+	if s := JobsSummary(preemptedJobs); s != "" {
```

Comment:
do  you want to log all this  at info level?  There could be  thousands of ids here I think?

Reply:
It does put a lot of stuff into the logs, but I don't think it's a large enough number to be an issue. The executor logs the ids of all jobs it renews leases for and this is typically a much smaller number of ids.

Reply:
I guess if we view ELK as the only viable log store then it's not completely unreasonable.  It's more just that generally I'd view this sort of verbose logging as something you'd put at debug level and specifically turn on to troubleshoot issues, rather than something you'd log at info proactively.

Reply:
> The executor logs the ids of all jobs it renews leases for and this is typically a much smaller number of ids.

This is also wrong!